site/content/entry/yuzu-fastmem/index.md

```diff
@@ -0,0 +1,118 @@
+
+&nbsp;
+
+{{< imgs
+	"./MMU.png|Illustration of an MMU (By Mdjango, Andrew S. Tanenbaum)"
+>}}
+
+In yuzu, we emulate the Switch's virtual address space for the running process. 
+But, that means for each time the game tries to read/write memory, which happens millions of times per frame, we need to basically "decode" where that Switch virtual address maps to our "emulated" Switch memory (RAM allocated for yuzu).
+This adds a lot of overhead when you consider how much the game reads/writes to memory.
+
+The idea behind fastmem here is to map the emulated Switch game's addressable memory on the host (Windows / Linux), at offset addresses. 
+The offset is a constant.
+
+## The Technical - HOW?
+
+On the Switch, we have 4GB of memory. 
+The hardware supports a virtual memory size of 48bits (just like x64); however, the Switch kernel limits this to only 39bit to save some memory for storing page tables.
+This results in a 512GB virtual address space (2^39) for use. (A page table is the data structure used to store the mapping between virtual addresses and physical addresses.)
+
+To emulate this, we allocated a HUGE page table, which had an offset for each memory page addressable in the virtual memory region of the Switch. Just this table was 1GB! 
+And whenever a game wanted to access the memory, we looked up this table, which resulted in roughly ~10 x64 CPU instructions and twice the memory latency: First read the table's value, then read the correct memory.
+
+Fastmem, however, uses the host MMU to rebuild the same 39bit virtual memory arena. 
+Here we're lucky that Nintendo limits the memory to 39bit. 
+We cannot allocate 48bit because that's the entire virtual address space on the host. 
+
+This cuts the effort to access the memory down to 3+1 x64 CPU instructions with a single memory latency.
+3 instructions to check if the pointer is smaller than 39 bit, and one memory load instruction. 
+We need this size check to ensure out-of-bounds memory access won't return any physical address.
+
+
+## The Challenges
+
+Implementing Fastmem required two major things on [Dynarmic](https://github.com/MerryMage/dynarmic/), our JIT recompiler. 
+
+**First:** A64 Fastmem support on Dynarmic.
+
+Dynarmic has had support for A32 Fastmem (see Citra Android) for a while now. 
+A64 Fastmem support was [originally worked on a year ago by MerryMage and vdwjeremy](https://github.com/MerryMage/dynarmic/pull/528) but it never got merged to the master branch. 
+
+[degasus](https://github.com/degasus/) worked with [MerryMage](https://github.com/MerryMage/) to clean up and get these changes merged to the master branch so that yuzu could benefit from this. [See here](https://github.com/MerryMage/dynarmic/pull/613)
```

Comment:
```suggestion
[degasus](https://github.com/degasus/) worked with [MerryMage](https://github.com/MerryMage/) to clean up and get [these changes merged](https://github.com/MerryMage/dynarmic/pull/613) to the master branch so that yuzu could benefit from this.
```